src/v/security/ephemeral_credential.h

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+#include "security/acl.h"
+#include "security/scram_credential.h"
+#include "security/types.h"
+#include "serde/envelope.h"
+
+#include <seastar/core/sstring.hh>
+
+#include <iosfwd>
+
+namespace security {
+
+class ephemeral_credential
+  : public serde::envelope<ephemeral_credential, serde::version<0>> {
+public:
+    using rpc_adl_exempt = std::true_type;
+
+    ephemeral_credential() noexcept = default;
+    ephemeral_credential(
+      acl_principal principal,
+      credential_user user,
+      credential_password password,
+      ss::sstring mechanism) noexcept
+      : _principal(std::move(principal))
+      , _user(std::move(user))
+      , _password(std::move(password))
+      , _mechanism(std::move(mechanism)) {}
+
+    const acl_principal& principal() const { return _principal; }
+    const credential_user& user() const { return _user; }
+    const credential_password& password() const { return _password; }
+    const ss::sstring& mechanism() const { return _mechanism; }
+
+    auto serde_fields() {
+        return std::tie(_principal, _user, _password, _mechanism);
+    }
+
+private:
+    friend bool
+    operator==(const ephemeral_credential&, const ephemeral_credential&)
+      = default;
+
+    friend std::ostream& operator<<(std::ostream&, const ephemeral_credential&);
+
+    acl_principal _principal;
+    credential_user _user;
```

Comment:
nit: Point of clarification, why have a credential_user when the acl_principal typically also contains the username?

Reply:
It's expected in the future that there will be additional `principal_types` such as `group`. This PR somewhat hijacks that mechanism to create an `ephemeral_user` type.

Reply:
This PR allows some decoupling of the `credential_user` from the `acl_principal`. Now there can be multiple `ephemeral_credentials`, each with a different `user` and `password`, that are associated with the `acl_principal`.